### DISCUSSION TITLE
Code formatting rules

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/grandorgue#920] [Developers]
This topic is for discussing the code formating rules.

They will be applyed automatically after #918

**Comment by oleg68** [comment]
Is suggest the following rules

1. 2-space indentation
2. Using spaces instead of tabs

**Reply by rousseldenis** [reply]
Can you try to follow this guide:

https://github.com/pocc/pre-commit-hooks

In your local repository, try to add the .pre-commit-config.yml file with proposed configuration.
Then, launch `pre-commit install`
And then, `pre-commit run -a`

It should reformat all files using that config. Add those changes in a commit. Then, modify a part of code without formatting, `git add`, then git commit ... and the magic happens. Pre-commit run the hooks, and reformat automatically. Relaunch `git add` and `git commit`. The code is committed and you don't have to bother about formatting yourself. Quite comfortable, I use it for years now.

**Reply by oleg68** [reply]
Thank you, @rousseldenis , but your answer is closer to #918 rather to this issue.

**Reply by larspalo** [reply]
I'm perfectly ok with just 2 space indentations, but then I do think that the brackets also should be placed on the starting line and not on it's own line. Ie:

~~~
someFunction(anything) {
  some code;
}
~~~

is better, in my opinion compared to:

~~~
someFunction(anything)
{
  some code;
}
~~~

which I feel makes more sense with tab indentation (and its usually greater "space").

**Reply by rousseldenis** [reply]
@larspalo I agree with that too.

FYI, I'm currently testing pre-commit hooks formatting (with default configs - we can adapt after) there : #946 

It reformats automatically like this :

![image](https://user-images.githubusercontent.com/19529533/148404201-72764f95-5a33-4621-8b4a-2ad1b7283bce.png)

**Reply by rousseldenis** [reply]
The aim is not to take care about formatting when coding. The git commit command will reformat like that automatically before real commit.

**Reply by oleg68** [reply]
> I do think that the brackets also should be placed on the starting line and not on it's own line. Ie:
> 
> ```
> someFunction(anything) {
>   some code;
> }
> ```
> 
> is better, in my opinion compared to:
> 
> ```
> someFunction(anything)
> {
>   some code;
> }
> ```

But the first style is more used with Java rather than with C/C++.

I'd leave the second style because most of C/C++ programmers follow it.

**Reply by rousseldenis** [reply]
> }\nBut the first style is more used with Java rather than with C/C++

It seems that google-style formatting for c++ is using the first one. See above

**Reply by larspalo** [reply]
@oleg68 I've seen it used with c++ too so that really doesn't matter. My reasoning is that the difference between the bracket and the indented line is visually greater as I propose when 2 blankspaces are used as indentation. If tabs are used the classical c/c++ way is fine because it's (usually) a greater distance between opening bracket and the first indented line. (with 2 space indention the difference diagonally will be only one space to the starting bracket if it's placed on its own line)

Anyway, even in the c++ core guidelines you'll find usage of both versions. Mainly functions with bracket on its own line, but for loops with the bracket at the end of the line... And that document has the c++ inventor/designer as author...

**Comment by larspalo** [comment]
Further general coding standard suggestions.

- We use mostly completely desciptive variable names (single letters reserved only for looping counters).
- Personally I prefer camelCaseNotation to the_underscore_notation. The only exception is for member variables that start with m_someName.
- Lines should not (!) be broken unless longer than 80 spaces, and lines longer than 132 characters "should" be broken up.
- When line is divided into multiple lines it should be done
1. At "complete" logical evaluation endings only
2. For each and every argument/parameter
3. With the standard indentation

**Reply by oleg68** [reply]
> Further general coding standard suggestions.
> 
> * We use mostly completely desciptive variable names (single letters reserved only for looping counters).

Agree. But usually it increases the line lengths. So I regard using a good and common-undestandable shortages (for example, `msg` instead of `message`, `config` instead of `configuration`) as an acceptable compromise/

> * Personally I prefer camelCaseNotation to the_underscore_notation. The only exception is for member variables that start with m_someName.

I'd declare the following notation:

- Class Names: cammel case starting with an upper letter
- Class member ordinar functions - cammel case starting with an upper letter
- Class member static functions - cammel case starting with a lower letter
- Class constants and enum members: uppercase with underscores
- Class variables: m_CammelCase and p_CammelCase (for pointers)
- local variables and parameters: cammelCase starting with a lower letter. The first letter `p` is preferable for pointers.
- static variables and static functions outside classes: lowercase with underscores

> * Lines should not (!) be broken unless longer than 80 spaces, and lines longer than 132 characters "should" be broken up.

I'd tell more strict: lines should never exceed 80 chars in length and should be broken at the maximal length before 80.

> * When line is divided into multiple lines it should be done
> 
> 1. At "complete" logical evaluation endings only
> 2. For each and every argument/parameter
> 3. With the standard indentation

So the first parameter also must appear at the new line:

```
  some_func(
     prm1,
     prm2,
     ...
     prmN
  );
```

instead of
 
```
  some_func (prm1,
                      prm2,
                      ...
                      prmN);
```

Although the second style is used more often in C/C++, it may contradict the requirements of complete and descriptive variable and function names: if the names are quite long and we use complex expressions then the room for parameters may be insufficient in the second style.

**Comment by larspalo** [comment]
> So I regard using a good and common-undestandable shortages (for example, msg instead of message, config instead of configuration) as an acceptable compromise

Most of the time this will be perfectly ok. But I think we should always be clear about what config we're dealing with in the current context. For instance it would be better to name a config variable as cmbCfg than just cfg... Even in the specific classes I'd prefer to have the variables named like audioSettings, deviceSettings, or midiConfig, or whatever more descriptive than just config or settings so it's always immediately obvious what kind of settings or configurations this variable carries even when read out of context.

> I'd tell more strict: lines should never exceed 80 chars in length and should be broken at the maximal length before 80.

Ok, but I'd tell it equally strict from the other side too. Lines that won't exceed 80 characters should not be broken either.

Other than that I agree with your suggestions, even if the Class/function/variable naming might take some time to get used to. It would anyway be of great benefit if we get this unified in the source code, I think.

**Reply by oleg68** [reply]
> Ok, but I'd tell it equally strict from the other side too. Lines that won't exceed 80 characters should not be broken either.

Agree

**Comment by oleg68** [comment]
I wont argue against the java-style of braces:

```
if (cond) {
}
```

@larspalo @rousseldenis  If it is acceptable, let us use them in this style.

**Reply by larspalo** [reply]
@oleg68 Sure!

**Comment by oleg68** [comment]
I've experienced with clang-format and found a configuration that satisfy almost all our needs:

```
﻿---
AlignAfterOpenBracket: AlwaysBreak
AlignOperands: 'false'
BinPackArguments: 'false'
BinPackParameters: 'false'
BreakBeforeBinaryOperators: All
ContinuationIndentWidth: '2'
...
```

The only thing it is not good for me is lacking of line break before closing parentheses.

```
  some_func(
    prm1,
    prm2)
```

The [``AlignAfterOpenBracket: BlockIndent`` option](https://reviews.llvm.org/D109557) would solve this problem, but it will be available only from the 14-th version of clang when the 13-th is the recent now.

If everyone agree, I'd ask @rousseldenis to add the `.clang-format` file with the content above, to rerun clang-format and to submit the changes to you PR #946

**Reply by larspalo** [reply]
@oleg68 @rousseldenis This seems fine to me! We can live with having the closing parenthesis at the end of the line instead of on its own until that feature is implemented in clang.

**Reply by rousseldenis** [reply]
Ok great, I can implement those configs.

My idea is to do one hook by one as the diff generated is quite big. Make the config, merge then the following one.

**Reply by oleg68** [reply]
> My idea is to do one hook by one as the diff generated is quite big. Make the config, merge then the following one.

It is not possible. The right strategy is the following

1. Make a separate branch
2. Add `.clang-format`
3. Run ``clang-format`` manually for reformatting the code at the first time
4. Commit all changes (`.clang-format` + all reformatted files) and make a PR
5. Afrer merging the PR rebase the github hook PR #946 on it.

**Reply by rousseldenis** [reply]
I didn't say the contrary : 😁

**Reply by rousseldenis** [reply]
I spoke about other hooks like uncrustify. Those ones can be added after clang if we want them

**Reply by oleg68** [reply]
Let start with clang-format only.

**Comment by larspalo** [comment]
@oleg68 @rousseldenis Sorry, but how do I get the pre-commit hook to do its work? Shouldn't that be done as I commit? Seems like it failed in my fork.

**Reply by rousseldenis** [reply]
Yes, but you have to do 'pre-commit install' once before. Either, after, you can run it manually with 'pre-commit run -a' to correct current code

**Reply by larspalo** [reply]
@rousseldenis Managed to get it installed, but it still fails with the code formatting. How much do I need to fix myself? Needless to say the real build of course works.

**Reply by larspalo** [reply]
Got it working by providing brackets {} for the if statement, seems like the pre-commit got confused when the single line (that was split due to line length) wasn't contained within them!
